### PR TITLE
JAMES-2586 Introduce PostgresExtension

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve https://github.com/linagora/james-project/issues/4894

### CODE REVIEW

**Comment by chibenwa** [comment]
Can we have build time statistics?

 - What is the overhead per test?
 - What is the overhead per test class ? From what I could see we do not recycle PG containers, and start a new one each time...

File: backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    private final PostgreSQLContainer postgresContainer;
+    private final PostgresModule postgresModule;
+    private PostgresqlConnectionFactory connectionFactory;
+    private PostgresExecutor postgresExecutor;
+    private PostgresTableManager postgresTableManager;
+
+    public PostgresExtension(PostgresModule postgresModule) {
+        this.postgresContainer = new PostgreSQLContainer<>(PostgresFixture.IMAGE)
```
Comment (vttranlina) [inline backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java:53]: Can we re-use `PostgresFixture.PG_CONTAINER.get()` ?

File: backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            .withUsername(PostgresFixture.Database.DB_USER)
+            .withPassword(PostgresFixture.Database.DB_PASSWORD)
+            .withLogConsumer(PostgresExtension::displayDockerLog)
+            .withCreateContainerCmdModifier(cmd -> cmd.withName("james-postgres-test-" + UUID.randomUUID()));
+
+        this.postgresModule = postgresModule;
+    }
+
+    public PostgresExtension() {
+        this.postgresContainer = new PostgreSQLContainer<>(PostgresFixture.IMAGE)
```
Comment (vttranlina) [inline backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java:64]: Reuse above constructor ? `this(PostgresModule.EMPTY_MODULE);`

**Comment by vttranlina** [comment]
Could you remove un-used `org.apache.james.backends.postgres.PostgresClusterExtension` ?

File: backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    @Override
+    public void beforeEach(ExtensionContext extensionContext) {
+        initTablesAndIndexes();
+    }
+
+    @Override
+    public void afterEach(ExtensionContext extensionContext) {
+        truncateTablesAndIndexes();
```
Comment (vttranlina) [inline backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java:113]: Should we DROP SCHEMA, and RE-CREATE SCHEMA?

// For table with enable RLS, we need drop POLICY (truncate table and index is not enough for this case)

Code ex: 
```java
Flux.usingWhen(postgresExtension.getConnection()
                    .map(c -> (PostgresqlConnection) c),
                connection -> Mono.from(connection.createStatement("DROP SCHEMA " + PostgresFixture.Database.SCHEMA + " CASCADE").execute())
                    .then(Mono.from(connection.createStatement("CREATE SCHEMA " + PostgresFixture.Database.SCHEMA).execute()))
                    .flatMap(PostgresqlResult::getRowsUpdated),
                Connection::close)
            .collectList()
            .block();
```

**Comment by quantranhong1999** [comment]
> Can we have build time statistics?

Reuse Postgres container across test classes fasten almost x3 the test runtime. Thank for you suggestion.

image

**Review by Arsnael — COMMENTED** [review]
Read it. Looks ok to me

**Comment by quantranhong1999** [comment]
rebased on top of the latest `PostgresTableManager`.

**Review by Arsnael — APPROVED** [review]

**Comment by quantranhong1999** [comment]
Green.

Well, I guess I have some CI luck (than somebody else) :p

**Review by hungphan227 — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

**Comment by Arsnael** [comment]
Please rebase

**Pull request merged**